// react-styleguidist/lib/client/rsg-components/slots/IsolateButton.js
import "core-js/modules/es.function.name";
import React from 'react';
import PropTypes from 'prop-types';
import { MdFullscreen, MdFullscreenExit } from 'react-icons/md';
import ToolbarButton from 'rsg-components/ToolbarButton';
import getUrl from '../../utils/getUrl';

var IsolateButton = function IsolateButton(_ref) {
  var name = _ref.name,
      example = _ref.example,
      isolated = _ref.isolated;
  var testID = example ? name + "-" + example + "-isolate-button" : name + "-isolate-button";
  return isolated ? React.createElement(ToolbarButton, {
    href: getUrl({
      anchor: true,
      slug: name.toLowerCase()
    }),
    title: "Show all components",
    testId: testID
  }, React.createElement(MdFullscreenExit, null)) : React.createElement(ToolbarButton, {
    href: getUrl({
      name: name,
      example: example,
      isolated: true
    }),
    title: "Open isolated",
    testId: testID
  }, React.createElement(MdFullscreen, null));
};

IsolateButton.propTypes = {
  name: PropTypes.string.isRequired,
  example: PropTypes.number,
  isolated: PropTypes.bool
};
export default IsolateButton;
