// react-styleguidist/lib/scripts/consts.js
"use strict";

exports.__esModule = true;
exports.DOCS_THIRDPARTIES = exports.DOCS_DOCUMENTING = exports.DOCS_WEBPACK = exports.DOCS_COMPONENTS = exports.DOCS_CONFIG = exports.BUGS = exports.HOMEPAGE = void 0;
const HOMEPAGE = 'https://react-styleguidist.js.org/';
exports.HOMEPAGE = HOMEPAGE;
const BUGS = 'https://github.com/styleguidist/react-styleguidist/issues';
exports.BUGS = BUGS;
const DOCS_CONFIG = 'https://react-styleguidist.js.org/docs/configuration.html';
exports.DOCS_CONFIG = DOCS_CONFIG;
const DOCS_COMPONENTS = 'https://react-styleguidist.js.org/docs/components.html';
exports.DOCS_COMPONENTS = DOCS_COMPONENTS;
const DOCS_WEBPACK = 'https://react-styleguidist.js.org/docs/webpack.html';
exports.DOCS_WEBPACK = DOCS_WEBPACK;
const DOCS_DOCUMENTING = 'https://react-styleguidist.js.org/docs/documenting.html';
exports.DOCS_DOCUMENTING = DOCS_DOCUMENTING;
const DOCS_THIRDPARTIES = 'https://react-styleguidist.js.org/docs/thirdparties.html';
exports.DOCS_THIRDPARTIES = DOCS_THIRDPARTIES;
